### PR TITLE
Resize avatars to displayed sizes

### LINKED ISSUE #513: Resize avatars to displayed sizes
Running a [PageSpeed Insight audit](https://web.dev/measure/?url=https%3A%2F%2Frailsdevs.com) on railsdevs highlights a few areas of improvement. This issue hopes to remedy the following:

> Avoid enormous network payloads Total size was 6,175 KiB

## A PR addressing this issue should:

- [ ] Use [Active Storage variants](https://api.rubyonrails.org/v6.1.0/classes/ActiveStorage/Variant.html) to resize avatars to appropriate sizes
- [ ] Ensure that 1x and 2x sizes work correctly for relevant devices

### PR DESCRIPTION
**Pull request by bradpurchase**
Closes #513.

- Uses ActiveStorage variants to resize avatars to their displayed sizes and updates the `AvatarComponent` component to make it easy to optionally display avatars as a variant.
- Adds `image_processing` to the Gemfile as this is required by ActiveStorage for resizing
- Adds `vips` to the Brewfile as this is a dependency of `image_processing` 

## Pull request checklist

- [x] My code contains tests covering the code I modified
- [x] I linted and tested the project with `bin/check`
- [x] I added significant changes and product updates to the [changelog](CHANGELOG.md)

### CODE REVIEW

**Comment by joemasilotti** [comment]
This is looking good!

A few questions:

1. Do we need all the changes to the Brewfile.lock? It looks like a bunch of packages got updated. I'd prefer to not do that in this PR if possible. And stick to only adding the vips-related stuff. 
2. Does this handle 2x screen resolutions? I don't see that anywhere. 
3. Do we need a backfill to resize all the existing images? Or will the first few page loads take a performance hit?

**Comment by bradpurchase** [comment]
Quick note on the first point: interestingly enough, it seems like using `brew bundle install --no-upgrade` will in fact still upgrade if homebrew determines it "needs to". See [this thread](https://github.com/Homebrew/homebrew-bundle/issues/632) 😕

**Comment by joemasilotti** [comment]
Awesome, thanks!

So we will serve 2x to everyone? If so, can you log an issue to add 1x and 2x via HTML to browsers?

And yes, let's create a script that will generate all variants of avatars that don't exist. Then I can backfill it before running the audit again.

**Comment by bradpurchase** [comment]
> So we will serve 2x to everyone? If so, can you log an issue to add 1x and 2x via HTML to browsers?

Yes. Do you mean specifying e.g. `@2x` on the image src in HTML? I can create an issue for that, if so.

> And yes, let's create a script that will generate all variants of avatars that don't exist. Then I can backfill it before running the audit again.

Sure, I'll get on that!

**Comment by joemasilotti** [comment]
Yes please! And sounds great, thanks.

**Comment by bradpurchase** [comment]
I've provided the backfill task, and created the issue [here](https://github.com/joemasilotti/railsdevs.com/issues/519). Let me know if there's anything I should add to that issue, or any other way I can help!

**Comment by joemasilotti** [comment]
Is it possible to only process variants for those that don't exist?

**Comment by bradpurchase** [comment]
> Is it possible to only process variants for those that don't exist?

Actually I believe `.processed` takes care of that, it will either process the variant if it doesn't exist, or return the processed variant if it does. See [here](https://edgeapi.rubyonrails.org/classes/ActiveStorage/Variant.html#method-i-processed).

**Comment by joemasilotti** [comment]
Beautiful!

I'm going to dig through the code one more time and do some local testing this weekend.

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -78,6 +78,11 @@ jobs:
         with:
           bundler-cache: true
 
+      - name: Install system deps
```
Comment (joemasilotti) [inline .github/workflows/ci.yml:81]:
```suggestion
      - name: Install system dependencies
```

File: .github/workflows/ci.yml [hunk 2]
```diff
@@ -78,6 +78,11 @@ jobs:
...
+        run: |
+          sudo apt-get -y update
+          sudo apt-get install -yqq libvips-dev
```
Comment (joemasilotti) [inline .github/workflows/ci.yml:84]: Is there no built in or third party Action to get us this? If not, is this the recommended way for CI? Genuine question.

File: app/components/avatar_component.rb [hunk 3]
```diff
@@ -1,15 +1,21 @@
...
 
-  attr_reader :avatarable, :image, :data
+  attr_reader :avatarable, :variant, :data
 
-  def initialize(avatarable:, classes: nil, data: {})
+  def initialize(avatarable:, variant: nil, classes: nil, data: {})
     @avatarable = avatarable
-    @image = avatarable&.avatar&.attached? ? avatarable.avatar : DEFAULT_AVATAR
+    @variant = variant
+    @image = image
```
Comment (joemasilotti) [inline app/components/avatar_component.rb:9]: I'm not sure what this line is doing.

File: test/components/avatar_component_test.rb [hunk 4]
```diff
@@ -14,6 +14,14 @@ class AvatarComponentTest < ViewComponent::TestCase
     assert_selector("img[src$='#{blob.filename}']")
   end
 
+  test "should render user avatar at specified variant" do
+    avatar_component = AvatarComponent.new(avatarable: @developer, variant: :thumb)
+    render_inline(avatar_component)
+
+    assert_equal true, avatar_component.image.is_a?(ActiveStorage::VariantWithRecord)
```
Comment (joemasilotti) [inline test/components/avatar_component_test.rb:21]:
```suggestion
    assert_instance_of ActiveStorage::VariantWithRecord, avatar_component.image
```

**Review by joemasilotti — CHANGES_REQUESTED** [review]
To sanity check, the avatars on the homepage and /developers are the same size as when viewing a developer profile. Right?

I made a few suggestions and questions in-line. But other than that this looks good to me!

I still need to do some manual testing.

**Comment by joemasilotti** [comment]
## PageSpeed score for `/developers`

### Mobile

* Before: 76, 79, 79
* After: 77, 82, 82

**Comment by joemasilotti** [comment]
The following changes were required to get vips working on Heroku:

1. Add the [apt buildpack](https://buildpack-registry.s3.amazonaws.com/buildpacks/heroku-community/apt.tgz) at index 0
2. Add the [heroku-buildpack-vips buildpack](https://github.com/brandoncc/heroku-buildpack-vips) as the last buildpack
3. Add the following to `Aptfile`:

```
libglib2.0-0
libglib2.0-dev
libpoppler-glib8
```

**Pull request merged**